### DISCUSSION TITLE
Interpolation of optional fields

### DISCUSSION THREAD

**Discussion by eonpatapon** [cuelang/cue#807] [Q&A]
Let say I have a few fields to interpolate in a string (some are optionals, some are not), my current solution is:

```
import "strings"

a: "foo"
b?: string

let R = strings.Join([
  a,
  if b != _|_ { b },
], "\n")
result: """
  stuff
  \(R)
  other stuff
"""
```

So the question is: is there a more simple way to do this ?

**Comment by verdverm** [comment]
Not sure if this is simpler, but another option is to use the `text/template` package instead of interpolation. I generally prefer `text/template` when

- the data is mildly complex
- the content starts getting long enough
- you want conditional content based on the values
- you want to loop over data

(self plug) https://github.com/hofstadter-io/hof is a tool for when you want to generate many files from a schema defined in Cue.

**Comment by mpvl** [comment]
If `a` and `b` were in a separate struct, you could use a comprehension to iterate over them. This will skip optional fields.

Also some of the proposed constructs in the query proposal will help here.

**Reply by eonpatapon** [reply]
> If `a` and `b` were in a separate struct, you could use a comprehension to iterate over them. This will skip optional fields.

Oh cool I didn't know that! Thanks

**Reply by seh** [reply]
> This will skip optional fields.

Do you mean optional _but non-concrete_ fields? Would it skip a populated value for `b`?

**Reply by mpvl** [reply]
When unified with a non-optional field, it will still iterate over it, even when not concrete.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/807.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
